### PR TITLE
Add support for the streaming dataflow API with checkpointing.

### PR DESCRIPTION
**Pull request by kl0u**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -94,6 +106,7 @@
 		<dependency>
 			<groupId>com.google.cloud.dataflow</groupId>
 			<artifactId>google-cloud-dataflow-java-sdk-all</artifactId>
+			<!--<version>[1.3]</version>-->
```
Comment (mxm) [inline pom.xml]: I think this can be removed.

File: pom.xml [hunk 2]
```diff
@@ -106,6 +119,7 @@
 		<dependency>
 			<groupId>com.google.cloud.dataflow</groupId>
 			<artifactId>google-cloud-dataflow-java-examples-all</artifactId>
+			<!--<version>[1.3]</version>-->
```
Comment (mxm) [inline pom.xml]: I think this can be removed.

File: pom.xml [hunk 3]
```diff
@@ -114,6 +128,28 @@
...
+			<artifactId>flink-connector-kafka</artifactId>
+			<version>${flink.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>org.mockito</groupId>
+			<artifactId>mockito-all</artifactId>
+			<version>1.9.5</version>
+			<scope>test</scope>
+		</dependency>
+		<!--<dependency>-->
```
Comment (mxm) [inline pom.xml]: I think this can be removed.

File: src/main/java/com/dataartisans/flink/dataflow/FlinkJobExecutionEnvironment.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
+package com.dataartisans.flink.dataflow;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/FlinkJobExecutionEnvironment.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/FlinkJobExecutionEnvironment.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.streaming.api.TimeCharacteristic;
+import org.apache.flink.streaming.api.environment.RemoteStreamEnvironment;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+// todo add tests for the streaming.
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/FlinkJobExecutionEnvironment.java]: TODOs stay in the code most of the times. It's better to address them directly or take note of them somewhere else, e.g. as an issue.

File: src/main/java/com/dataartisans/flink/dataflow/examples/WordCount.java [hunk 6]
```diff
@@ -35,15 +35,16 @@
 	 */
 	public interface Options extends PipelineOptions, FlinkPipelineOptions {
 		@Description("Path of the file to read from")
-		@Default.String("gs://dataflow-samples/shakespeare/kinglear.txt")
+		@Default.String("/Users/kkloudas/Desktop/randomtext_lorem_p.txt")
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/examples/WordCount.java]: Please undo the change.

File: src/main/java/com/dataartisans/flink/dataflow/examples/WordCount.java [hunk 7]
```diff
@@ -35,15 +35,16 @@
...
 		String getInput();
 		void setInput(String value);
 
 		@Description("Path of the file to write to")
+		@Default.String("/Users/kkloudas/Desktop/out.txt")
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/examples/WordCount.java]: Please undo the change.

File: src/main/java/com/dataartisans/flink/dataflow/examples/streaming/AutoComplete.java [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+   * <p> Inherits standard Dataflow configuration options.
+   */
+  private static interface Options extends WindowedWordCount.StreamingWordCountOptions {
+    @Description("Whether to use the recursive algorithm")
+    @Default.Boolean(true)
+    Boolean getRecursive();
+    void setRecursive(Boolean value);
+  }
+
+  public static void main(String[] args) throws IOException {
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/examples/streaming/AutoComplete.java]: Does this example work out of the box?

File: src/main/java/com/dataartisans/flink/dataflow/examples/streaming/JoinExamples.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+			String key = line.split("\\s")[0];
+			c.output(KV.of(key, line));
+		}
+	}
+
+	private static interface Options extends WindowedWordCount.StreamingWordCountOptions {
+
+	}
+
+	public static void main(String[] args) throws Exception {
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/examples/streaming/JoinExamples.java]: This this work out of the box?

File: src/main/java/com/dataartisans/flink/dataflow/examples/streaming/KafkaWindowedWordCountExample.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
+package com.dataartisans.flink.dataflow.examples.streaming;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/examples/streaming/KafkaWindowedWordCountExample.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/examples/streaming/WindowedWordCount.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
+package com.dataartisans.flink.dataflow.examples.streaming;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/examples/streaming/WindowedWordCount.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkBatchPipelineTranslator.java [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+import com.google.cloud.dataflow.sdk.options.PipelineOptions;
+import com.google.cloud.dataflow.sdk.runners.TransformTreeNode;
+import com.google.cloud.dataflow.sdk.transforms.AppliedPTransform;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.join.CoGroupByKey;
+import com.google.cloud.dataflow.sdk.values.PValue;
+import org.apache.flink.api.java.ExecutionEnvironment;
+
+/**
+ * FlinkPipelineTranslator knows how to translate Pipeline objects into Flink Jobs.
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkBatchPipelineTranslator.java]: Update class name here.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkPipelineTranslator.java [hunk 13]
```diff
@@ -1,168 +1,13 @@
-/*
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkPipelineTranslator.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkPipelineTranslator.java [hunk 14]
```diff
@@ -1,168 +1,13 @@
...
-import com.google.cloud.dataflow.sdk.transforms.PTransform;
-import com.google.cloud.dataflow.sdk.transforms.join.CoGroupByKey;
-import com.google.cloud.dataflow.sdk.values.PValue;
-import org.apache.flink.api.java.ExecutionEnvironment;
 
 /**
- * FlinkPipelineTranslator knows how to translate Pipeline objects into Flink Jobs.
- *
- * This is based on {@link com.google.cloud.dataflow.sdk.runners.DataflowPipelineTranslator}
+ * Created by kkloudas on 10/23/15.
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkPipelineTranslator.java]: Please remove this line.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingPipelineTranslator.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
+package com.dataartisans.flink.dataflow.translation;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingPipelineTranslator.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingPipelineTranslator.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+
+import com.google.cloud.dataflow.sdk.options.PipelineOptions;
+import com.google.cloud.dataflow.sdk.runners.TransformTreeNode;
+import com.google.cloud.dataflow.sdk.transforms.AppliedPTransform;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.join.CoGroupByKey;
+import com.google.cloud.dataflow.sdk.values.PValue;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+
+// todo see merging with the batch translator.
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingPipelineTranslator.java]: I suppose it makes sense to keep this separate. We could, however, put some of the code in the super class `FlinkPipelineTranslator`. Not high priority.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingPipelineTranslator.java [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+		System.out.println(genSpaces(this.depth) + "enterCompositeTransform- " + formatNodeName(node));
+
+		PTransform<?, ?> transform = node.getTransform();
+		if (transform != null && currentCompositeTransform == null) {
+
+			StreamTransformTranslator<?> translator = FlinkStreamingTransformTranslators.getTranslator(transform);
+			if (translator != null) {
+				currentCompositeTransform = transform;
+				if (transform instanceof CoGroupByKey && node.getInput().expand().size() != 2) {
+					// we can only optimize CoGroupByKey for input size 2
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingPipelineTranslator.java]: Do we make this optimization also for streaming?

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTransformTranslators.java [hunk 18]
```diff
@@ -0,0 +1,342 @@
+package com.dataartisans.flink.dataflow.translation;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTransformTranslators.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTransformTranslators.java [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+import org.apache.flink.streaming.api.datastream.*;
+import org.apache.flink.util.Collector;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+/**
+ * todo integrate type checks like the following:
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTransformTranslators.java]: Could you address the TODO or put it in an issue?

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTranslationContext.java [hunk 20]
```diff
@@ -0,0 +1,136 @@
+package com.dataartisans.flink.dataflow.translation;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTranslationContext.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTranslationContext.java [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.typeutils.GenericTypeInfo;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by kkloudas on 10/19/15.
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTranslationContext.java]: Please remove.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTranslationContext.java [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+	public PipelineOptions getPipelineOptions() {
+		return options;
+	}
+
+	@SuppressWarnings("unchecked")
+	public <T> DataStream<T> getInputDataStream(PValue value) {
+		return (DataStream<T>) dataStreams.get(value);
+	}
+
+	public void setWindowingFn(PValue value, WindowFn<?,?> windowFn) {
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTranslationContext.java]: Unused function.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkAbstractParDoWrapper.java [hunk 23]
```diff
@@ -0,0 +1,259 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkAbstractParDoWrapper.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkAbstractParDoWrapper.java [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+			} catch (Exception e) {
+				Throwables.propagateIfInstanceOf(e, UserCodeException.class);
+				throw new UserCodeException(e);
+			}
+		}
+
+		return WindowedValue.of(output, timestamp, windows, pane);
+	}
+
+	///////////			ABSTRACT METHODS TO BE IMPLEMENTED BY SUBCLASSES			/////////////////
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkAbstractParDoWrapper.java]: Can be removed?

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkGroupAlsoByWindowWrapper.java [hunk 25]
```diff
@@ -0,0 +1,586 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkGroupAlsoByWindowWrapper.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkGroupByKeyWrapper.java [hunk 26]
```diff
@@ -0,0 +1,41 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkGroupByKeyWrapper.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkParDoBoundMultiWrapper.java [hunk 27]
```diff
@@ -0,0 +1,57 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkParDoBoundMultiWrapper.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkParDoBoundWrapper.java [hunk 28]
```diff
@@ -0,0 +1,74 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkParDoBoundWrapper.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/Utils.java [hunk 29]
```diff
@@ -0,0 +1,62 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/Utils.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/io/UnboundedFlinkSource.java [hunk 30]
```diff
@@ -0,0 +1,61 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming.io;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/io/UnboundedFlinkSource.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/io/UnboundedSocketSource.java [hunk 31]
```diff
@@ -0,0 +1,211 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming.io;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/io/UnboundedSocketSource.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/io/UnboundedSocketSource.java [hunk 32]
```diff
@@ -0,0 +1,211 @@
...
+
+	@Override
+	public UnboundedReader<String> createReader(PipelineOptions options, @Nullable C checkpointMark) {
+		return new UnboundedSocketReader(this);
+	}
+
+	@Nullable
+	@Override
+	public Coder getCheckpointMarkCoder() {
+		return null; // todo check this
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/io/UnboundedSocketSource.java]: We should check this.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/io/UnboundedSourceWrapper.java [hunk 33]
```diff
@@ -0,0 +1,105 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming.io;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/io/UnboundedSourceWrapper.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/AbstractFlinkTimerInternals.java [hunk 34]
```diff
@@ -0,0 +1,124 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming.state;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/AbstractFlinkTimerInternals.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/FlinkStateInternals.java [hunk 35]
```diff
@@ -0,0 +1,521 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming.state;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/FlinkStateInternals.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/FlinkStateInternals.java [hunk 36]
```diff
@@ -0,0 +1,521 @@
...
+		}
+
+		@Override
+		public void persistState(StateCheckpointWriter checkpointBuilder) throws IOException {
+			if (value != null) {
+
+				// serialize the coder.
+				byte[] coder = InstantiationUtil.serializeObject(elemCoder);
+
+// 				TODO: 12/19/15 check why this does not work.
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/FlinkStateInternals.java]: Did you check this?

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateCheckpointReader.java [hunk 37]
```diff
@@ -0,0 +1,74 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming.state;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateCheckpointReader.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateCheckpointUtils.java [hunk 38]
```diff
@@ -0,0 +1,137 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming.state;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateCheckpointUtils.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateCheckpointWriter.java [hunk 39]
```diff
@@ -0,0 +1,112 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming.state;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateCheckpointWriter.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateType.java [hunk 40]
```diff
@@ -0,0 +1,56 @@
+package com.dataartisans.flink.dataflow.translation.wrappers.streaming.state;
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateType.java]: Missing license header.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateType.java [hunk 41]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.io.IOException;
+
+public enum StateType {
+
+	VALUE,
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateType.java]: How about `VALUE(0)`, `WATERMARK(1)`,...

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateType.java [hunk 42]
```diff
@@ -0,0 +1,56 @@
...
+	ACCUMULATOR;
+
+	public static void serialize(StateType type, StateCheckpointWriter output) throws IOException {
+		if (output == null) {
+			throw new IllegalArgumentException("Cannot write to a null output.");
+		}
+
+		switch (type) {
+			case VALUE:
+				output.writeByte((byte) 0);
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/state/StateType.java]: You could assign numbers to the possible values of the Enum in the constructor. That way you don't loose track in the serialize/deserialize methods.

File: src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java [hunk 43]
```diff
@@ -0,0 +1,517 @@
+package com.dataartisans.flink.dataflow.streaming;
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java]: Missing license header.

File: src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java [hunk 44]
```diff
@@ -0,0 +1,517 @@
...
+						inputCoder,
+						combiner.<String>asKeyedFn());
+
+		OneInputStreamOperatorTestHarness<WindowedValue<KV<String, Integer>>, WindowedValue<KV<String, Integer>>> testHarness =
+				new OneInputStreamOperatorTestHarness<>(gbwOperaror);
+		testHarness.open();
+
+		/**
+		 * If it fires once then the window is considered closed. If not, then if elements<count -> output
+		 * There is some cleanup going on that I have to find and check.
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java]: Do you still have to check? :)

File: src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java [hunk 45]
```diff
@@ -0,0 +1,517 @@
...
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java]: Do you still have to check? :)

File: src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java [hunk 46]
```diff
@@ -0,0 +1,517 @@
...
+				new Instant(initialTime + 19500), null, PaneInfo.createPane(true, true, PaneInfo.Timing.ON_TIME)), initialTime));
+
+		expectedOutput.add(new Watermark(initialTime + 20000));
+		TestHarnessUtil.assertOutputEqualsSorted("Output was not correct.", expectedOutput, testHarness.getOutput(), new ResultSortComparator());
+
+		testHarness.close();
+	}
+
+	@Test
+	public void testCompoundAccumilatingPanesProgram() throws Exception {
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java]: typo: Accumulating

File: src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java [hunk 47]
```diff
@@ -0,0 +1,517 @@
...
+						inputCoder,
+						combiner.<String>asKeyedFn());
+
+		OneInputStreamOperatorTestHarness<WindowedValue<KV<String, Integer>>, WindowedValue<KV<String, Integer>>> testHarness =
+				new OneInputStreamOperatorTestHarness<>(gbwOperaror);
+		testHarness.open();
+
+		/**
+		 * If it fires once then the window is considered closed. If not, then if elements<count -> output
+		 * There is some cleanup going on that I have to find and check.
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java]: Do you still have to check? :)

File: src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java [hunk 48]
```diff
@@ -0,0 +1,517 @@
...
+		testHarness.processElement(new StreamRecord<>(makeWindowedValue(strategy, KV.of("key1", 1), new Instant(initialTime + 19500), null, PaneInfo.NO_FIRING), initialTime + 20));
+		testHarness.processElement(new StreamRecord<>(makeWindowedValue(strategy, KV.of("key1", 1), new Instant(initialTime + 19500), null, PaneInfo.NO_FIRING), initialTime + 20));
+		testHarness.processElement(new StreamRecord<>(makeWindowedValue(strategy, KV.of("key1", 1), new Instant(initialTime + 19500), null, PaneInfo.NO_FIRING), initialTime + 20));
+		testHarness.processElement(new StreamRecord<>(makeWindowedValue(strategy, KV.of("key1", 1), new Instant(initialTime + 19500), null, PaneInfo.NO_FIRING), initialTime + 20));
+		testHarness.processElement(new StreamRecord<>(makeWindowedValue(strategy, KV.of("key1", 1), new Instant(initialTime + 19500), null, PaneInfo.NO_FIRING), initialTime + 20));
+		testHarness.processElement(new StreamRecord<>(makeWindowedValue(strategy, KV.of("key1", 1), new Instant(initialTime + 19500), null, PaneInfo.NO_FIRING), initialTime + 20));
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java]: This is the only testing data. It seems like this could be automated to generate more keys and interleavings.

File: src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java [hunk 49]
```diff
@@ -0,0 +1,517 @@
...
+
+	private static class ResultSortComparator implements Comparator<Object> {
+		@Override
+		public int compare(Object o1, Object o2) {
+			if (o1 instanceof Watermark && o2 instanceof Watermark) {
+				Watermark w1 = (Watermark) o1;
+				Watermark w2 = (Watermark) o2;
+				return (int) (w1.getTimestamp() - w2.getTimestamp());
+			} else if (o1 instanceof Watermark || o2 instanceof Watermark) {
+				return 1;
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java]: This check may not produce a deterministic ordering.

File: src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java [hunk 50]
```diff
@@ -0,0 +1,517 @@
...
+					BoundedWindow windowA = (BoundedWindow) windowsA.iterator().next();
+					BoundedWindow windowB = (BoundedWindow) windowsB.iterator().next();
+					comparison = Long.compare(windowA.maxTimestamp().getMillis(), windowB.maxTimestamp().getMillis());
+				}
+				return comparison;
+			}
+		}
+	}
+
+	private <T> WindowedValue<T> makeWindowedValue(WindowingStrategy strategy,
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java]: formatting

File: src/test/java/com/dataartisans/flink/dataflow/streaming/StateSerializationTest.java [hunk 51]
```diff
@@ -0,0 +1,242 @@
+package com.dataartisans.flink.dataflow.streaming;
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/StateSerializationTest.java]: Missing license header.

File: src/test/java/com/dataartisans/flink/dataflow/streaming/StateSerializationTest.java [hunk 52]
```diff
@@ -0,0 +1,242 @@
...
+		}
+		return comp;
+	}
+
+	private void storeState(StateBackend.CheckpointStateOutputView out) throws Exception {
+		StateCheckpointWriter checkpointBuilder = StateCheckpointWriter.create(out);
+		Coder<String> keyCoder = StringUtf8Coder.of();
+
+		// checkpoint the timers
+		StateCheckpointUtils.encodeTimers(activeTimers, checkpointBuilder,keyCoder);//encodeTimers(out, keyCoder);
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/StateSerializationTest.java]: Left over comment.

File: src/test/java/com/dataartisans/flink/dataflow/streaming/StateSerializationTest.java [hunk 53]
```diff
@@ -0,0 +1,242 @@
...
+	private FlinkStateInternals<String> createState(String key) throws CannotProvideCoderException {
+		return new FlinkStateInternals<>(
+				key,
+				StringUtf8Coder.of(),
+				IntervalWindow.getCoder(),
+				combiner.asKeyedFn());
+	}
+
+	@Test
+	public void test() throws Exception {
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/StateSerializationTest.java]: Just my personal opinion. This test could have been broken into multiple tests. The init and teardown code could have been put in `@Before` and `@After` methods. 

To me, it seems much easier for debugging to do an `assertTrue` instead of ANDing all the checks.

File: src/test/java/com/dataartisans/flink/dataflow/util/JoinExamples.java [hunk 54]
```diff
@@ -76,8 +76,8 @@ public void processElement(ProcessContext c) {
 						KV<String, CoGbkResult> e = c.element();
 						CoGbkResult val = e.getValue();
 						String countryCode = e.getKey();
-						String countryName;
-						countryName = e.getValue().getOnly(countryInfoTag);
+						String countryName = "none";
```
Comment (mxm) [inline src/test/java/com/dataartisans/flink/dataflow/util/JoinExamples.java]: Are these changes necessary?

**Comment by mxm** [comment]
Thanks for the pull request @kl0u. Looks very good at first sight. I definitely want to merge this after trying it out. There are already some users that want to try it out as well :) 

Just saw that there are custom sources but no custom sinks but we can address this in a followup.

**Comment by mxm** [comment]
Also, it would be nice to add some documentation on how the Dataflow streaming model is executed on top of Flink and what the features and limitations are. As far as I see, we support pretty much all the features because we reuse Dataflow's windowing and triggering logic.

File: src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTransformTranslators.java [hunk 55]
```diff
@@ -0,0 +1,342 @@
...
+	// here you can find all the available translators.
+	static {
+		GOOG_TO_FLINK_TRANSLATORS.put(Read.Unbounded.class, new UnboundedReadSourceTranslator());
+		GOOG_TO_FLINK_TRANSLATORS.put(ParDo.Bound.class, new ParDoBoundStreamingTranslator());
+		GOOG_TO_FLINK_TRANSLATORS.put(TextIO.Write.Bound.class, new TextIOWriteBoundStreamingTranslator());
+		GOOG_TO_FLINK_TRANSLATORS.put(Window.Bound.class, new WindowBoundTranslator());
+		GOOG_TO_FLINK_TRANSLATORS.put(GroupByKey.class, new GroupByKeyTranslator());
+		GOOG_TO_FLINK_TRANSLATORS.put(Combine.PerKey.class, new CombinePerKeyTranslator());
+		GOOG_TO_FLINK_TRANSLATORS.put(Flatten.FlattenPCollectionList.class, new FlattenPCollectionTranslator());
+		GOOG_TO_FLINK_TRANSLATORS.put(ParDo.BoundMulti.class, new ParDoBoundMultiStreamingTranslator());
```
Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/FlinkStreamingTransformTranslators.java]: Can we rename these to `TRANSLATORS`? :)

File: src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java (hunk 44 above)
Comment (kl0u) [inline src/test/java/com/dataartisans/flink/dataflow/streaming/GroupAlsoByWindowTest.java]: Yes, so I will remove it from comments and I will add it to the issues.

File: src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkAbstractParDoWrapper.java (hunk 24 above)
Comment (kl0u) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkAbstractParDoWrapper.java]: I think this comment makes the code clearer to someone who reads it, as these are the methods to be implemented by subclasses.

Comment (mxm) [inline src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkAbstractParDoWrapper.java]: Fair enough. Let's keep it then.

**Comment by mxm** [comment]
Merged in b5c2d7e010ad3eaef5b2253af151ce5f47ab42fe.

**Comment by kl0u** [comment]
Perfect! Thanks a lot @mxm for the help and the comments!
